src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs

```diff
@@ -555,6 +712,205 @@ protected virtual unsafe IntPtr AppWndProc(IntPtr hWnd, uint msg, IntPtr wParam,
+            if (historyCount > 1)
+            {
+                return new Lazy<IReadOnlyList<RawPointerPoint>>(() =>
+                {
+                    s_intermediatePointsPooledList.Clear();
+                    s_intermediatePointsPooledList.Capacity = historyCount;
+
+                    // Pointers in history are ordered from newest to oldest, so we need to reverse iteration.
+                    // Also we skip the newest pointer, because original event arguments already contains it.
+
+                    if (info.pointerType == PointerInputType.PT_TOUCH)
+                    {
+                        if (GetPointerTouchInfoHistory(info.pointerId, ref historyCount, s_historyTouchInfos))
+                        {
+                            for (int i = historyCount - 1; i >= 1; i--)
+                            {
+                                var historyTouchInfo = s_historyTouchInfos[i];
+                                s_intermediatePointsPooledList.Add(CreateRawPointerPoint(historyTouchInfo));
+                            }
+                        }
+                    }
+                    else if (info.pointerType == PointerInputType.PT_PEN)
+                    {
+                        if (GetPointerPenInfoHistory(info.pointerId, ref historyCount, s_historyPenInfos))
+                        {
+                            uint timestamp = 0;
+                            for (int i = historyCount - 1; i >= 1; i--)
+                            {
+                                var historyPenInfo = s_historyPenInfos[i];
+                                s_intermediatePointsPooledList.Add(CreateRawPointerPoint(historyPenInfo));
+                            }
+                        }
+                    }
+                    else
+                    {
+                        // Currently Windows does not return history info for mouse input, but we handle it just for case.
+                        if (GetPointerInfoHistory(info.pointerId, ref historyCount, s_historyInfos))
+                        {
+                            for (int i = historyCount - 1; i >= 1; i--)
+                            {
+                                var historyInfo = s_historyInfos[i];
+                                s_intermediatePointsPooledList.Add(CreateRawPointerPoint(historyInfo));
+                            }
+                        }
+                    }
+                    return s_intermediatePointsPooledList;
+                });
+            }
+
+            return null;
+        }
+
+        private RawPointerEventArgs CreatePointerArgs(IInputDevice device, ulong timestamp, RawPointerEventType eventType, RawPointerPoint point, RawInputModifiers modifiers, uint rawPointerId)
+        {
+            return device is TouchDevice
+                ? new RawTouchEventArgs(device, timestamp, _owner, eventType, point, modifiers, rawPointerId)
```

Comment:
RawTouchEventArgs is not needed anymore, since RawPointerEventArgs has all required information. But I left it for compatibility.